lib/agendaitem-order.js

```diff
@@ -0,0 +1,95 @@
+import { query, update, sparqlEscapeUri, sparqlEscapeInt } from 'mu';
+import { reduceResultSet } from './utils';
+
+async function getRelatedAgendaitems(agenda, agendaitemType) {
+  const queryString = `PREFIX dct: <http://purl.org/dc/terms/>
+PREFIX mandaat: <http://data.vlaanderen.be/ns/mandaat#>
+PREFIX schema: <http://schema.org/>
+PREFIX prov: <http://www.w3.org/ns/prov#>
+PREFIX besluitvorming: <https://data.vlaanderen.be/ns/besluitvorming#>
+PREFIX ext: <http://mu.semte.ch/vocabularies/ext/>
+
+SELECT DISTINCT ?agendaitem ?agendaitemCreated ?agendaitemPosition ?mandateePriority
+WHERE {
+  ${sparqlEscapeUri(agenda)} dct:hasPart ?agendaitem .
+  ?agendaitem schema:position ?agendaitemPosition ;
+              dct:created ?agendaitemCreated ;
+              dct:type ${sparqlEscapeUri(agendaitemType)} .
+  FILTER NOT EXISTS {
+    ?agendaitem prov:wasRevisionOf ?olderAgendaitem .
+  }
+  ?agendaActivity besluitvorming:genereertAgendapunt ?agendaitem ;
+    besluitvorming:vindtPlaatsTijdens ?subcase .
+  OPTIONAL {
+    ?subcase ext:heeftBevoegde ?mandatee .
+    ?mandatee mandaat:rangorde ?mandateePriority .
+  }
+}`;
+  return await query(queryString);
+}
+
+async function applyAgendaitemsOrder(orders) {
+  const queryString = `PREFIX schema: <http://schema.org/>
+
+DELETE {
+  ?agendaitem schema:position ?oldPosition .
+}
+INSERT {
+  ?agendaitem schema:position ?newPosition .
+}
+WHERE {
+  VALUES (?agendaitem ?oldPosition ?newPosition) {
+    ${orders
+      .map(
+        ({ uri, oldPosition, newPosition }) =>
+          `(${sparqlEscapeUri(uri)} ${sparqlEscapeInt(oldPosition)} ${sparqlEscapeInt(newPosition)})`)
+      .join('\n    ')}
+  }
+}`;
+  await update(queryString);
+}
+
+async function reorderAgendaitems(agenda, agendaitemType) {
+  const agendaitems = reduceResultSet(await getRelatedAgendaitems(agenda, agendaitemType));
+
+  // We sort on the mandatee priority. To do this, we simply cast the array of
+  // priorities to a string: [1, 2, 3] → "1,2,3" and we just use string
+  // comparisons to ensure we have a lexicographical sort. E.g. given the
+  // following list of priorities: [ [1, 2], [1], [2, 4], [3], [2, 3] ]
+  // the sorted list will be: [ [1], [1, 2], [2, 3], [2, 4], [3] ]
+  //
+  // If multiple items have the same priority list, we sort them based on their
+  // existing position. This way manual reorders inside a priority group are
+  // kept when submitting a new agendaitem.
+  const sortedAgendaitems = agendaitems.sort(
+    (a1, a2) => {
+      const priority1 = Array.isArray(a1.mandateePriority) ? a1.mandateePriority.sort().toString() : String(a1.mandateePriority);
+      const priority2 = Array.isArray(a2.mandateePriority) ? a2.mandateePriority.sort().toString() : String(a2.mandateePriority);
+      return priority1 === priority2
+        ? a1.agendaitemPosition - a2.agendaitemPosition
+        : priority1 < priority2
+          ? -1
+          : 1
+    });
```

Comment:
does this sorting also move the approval item?
If present, that should stay as the first agendaitem.

Reply:
The query to fetch the agendaitems of the latest agenda has the following bit: 
```sparql
  ?agendaActivity besluitvorming:genereertAgendapunt ?agendaitem ;
    besluitvorming:vindtPlaatsTijdens ?subcase .
```
https://github.com/kanselarij-vlaanderen/agenda-submission-service/pull/4/files#diff-5b6ddf4aacad6a5d676d764ced5b8b1a51d3ce8a3df4d7523b8fe7c450a6c983R21-R22

The approval shouldn't ever be found because it has no agenda activity.

Reply:
ah yes. this paired with the math.min should work.
The only time this would not work is when the approval item is not the first agendaitem.
It shouldn't really happen afaik but there are 2 old agendas in PROD where this is the case.
We could ask the question if this will ever happen again.
But if it does, the approval item will have a duplicate position after this code since we don't take it into account?

Reply:
Yes, just tested, and if the approval is moved to position 2, and then another agendaitem is added, you get 2 agendaitems with position 2.

Now, since moving the approval is clearly a user error, and it is very easily fixed by simply moving it back in place (or even to some other place on the agenda), I'm inclined to merge this as such, unless you really insist to have this fixed @ValenberghsSven ?